sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -102,17 +103,38 @@ int pollSocketUntilEventOrTimeout(
     interval = timeout;
   }
   int result = 0;
-  for (long counter = 0; counter < timeout && result == 0; counter = counter + interval)
+  bool poll_interrupted = false; // deadline and now are only valid if poll_interrupted
+  long deadline{};
+  long now{};
+  for (long counter = 0; (poll_interrupted ? now < deadline : counter < timeout) && result == 0;
```

Comment:
This for loop seems *really* awkward to me.

Essentially what seems to be happening here is that we start by iterating over the loop terminating when we have iterated for `timeout / interval` times.

However, if a signal is received, the loop termination condition changes to be "however many iterations are needed until the deadline is expired".

IMHO the loop termination condition should simply be whenever the deadline is exceeded.

One possible modification is to create a std::chrono::steady_clock variable named `deadline` and then change the loop from a for loop to something like:
```c++
long deadline = std::chrono::steady_clock::now() + std::chrono::duration<std::chrono::milliseconds>(timeout);
while (now < deadline)
{
  context.ThrowIfCancelled();
#if defined(AZ_PLATFORM_POSIX)
  result = poll(&poller, 1, interval);
  if (result < 0 && errno == EINTR)
  {
      continue;
  }
#elif defined(AZ_PLATFORM_WINDOWS)
  result = WSAPoll(&poller, 1, interval);
#endif
  if (result != 0)
  {
    return result;
  }
  now = std::chrono::steady_clock::now();
}
return 0;
```

With a structure like that, you don't need to do any special handling for the EINTR case - you just ignore EINTR errors being returned from poll, and then the rest of the code "just works".

To be clear: this is just my scribbling code into the github editor. you can structure the code as you feel appropriate. But I feel strongly that there should just be a single termination condition for the loop rather than varying response conditions.

Reply:
The comment indicates the motivation for the extra complexity, which is to avoid an extra syscall (to read the clock) in the common case.  Perhaps this is an unnecessary optimization but seems a shame to add overhead to what's got to be one of the most frequently executed codepaths in the sdk.  Even a single extra syscall in a poll/read or write loop can be significant in bulk data transfer, and the case where we actually need the clock here is quite rare.

Reply:
My primary concern here is the added complexity of the loop termination conditional. There are functionally two separate termination cases, depending on whether a EINTR was thrown or not.

Also remember that this is polling a network socket. This function is almost never going to be CPU bound on any platform. You're looking at a *minimum* round trip time to the server measured in tens of milliseconds, and a polling interval of 1 second. I'm hard pressed to see how reading the steady clock is going to be a performance bottleneck (on Windows, reading the steady clock is literally reading 8 bytes of data from a piece of shared memory, I'm not sure how the corresponding functionalty works on Linux, but I'm pretty sure it's comparably fast).

IMHO reducing code complexity almost always trumps performance, unless you have evidence that the function is a performance bottleneck. And I'm hard pressed to see how this function is a performance bottleneck. 

Larry


PS: If you are *truly* worried about CPU consumption inside this loop, if you should look at context.ThrowIfCancelled(), it also calls std::chrono::system_clock::now() in addition to calling GetDeadline() which loops through a stack of context objects looking for a context object with a deadline. So you can return to the original number of calls to `std::chrono::steady_clock::now()` by calling `context.GetDeadLine()` to retrieve the context deadline outside the loop and then setting the `deadline` variable to the min of deadline from the parameter and deadline from the context.

Reply:
Fair enough, especially as the branches aren't easily tested.

Reply:
> This has exceeded my time box, so I'll leave it in your hands from here on @LarryOsterman.

That's completely fair. The fix for the CI issue is straightforward, but annoying. 

Basically at the top of the file, you need to move the NOMINMAX goo:
```c++
#include "azure/core/platform.hpp"
#if defined(AZ_PLATFORM_WINDOWS)
#if !defined(NOMINMAX)
#define NOMINMAX
#endif
#endif

#include "azure/core/http/curl_transport.hpp"
```

because the Windows headers are profoundly broken.

I understand your timebox limits and appreciate them, but I'd *LOVE* it if you could get contributor credit for this change (since you did ALL the work here and it's a good improvement to the SDK).

I'm going to let you make the call about whether you make this modification to your PR, or if you want me to drive it forward.

If you don't want to invest any more time in this PR(and I can understand that), I see two ways of moving forward with this:
 1. I can clone your fork and make the change there. That way this pull request completes with your PR.
 2. I can integrate your changes in my PR. The bad news is that I don't know how to give you credit.

Let me know how you want to proceed.

Reply:
> The bad news is that I don't know how to give you credit.

Fwiw, at the very least, the changelog entry for the fix will have the acknowledgement of the contribution. :)

For example:
https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/CHANGELOG.md#acknowledgments